pkg/controller/trainingruntime_controller.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2025 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"iter"
+	"slices"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	trainer "github.com/kubeflow/trainer/pkg/apis/trainer/v1alpha1"
+	"github.com/kubeflow/trainer/pkg/constants"
+	idxer "github.com/kubeflow/trainer/pkg/runtime/indexer"
+	"github.com/kubeflow/trainer/pkg/trainjob"
+)
+
+const (
+	updateChBuffer = 10
+)
+
+type TrainingRuntimeReconciler struct {
+	log                      logr.Logger
+	client                   client.Client
+	recorder                 record.EventRecorder
+	nonRuntimeObjectUpdateCh chan event.TypedGenericEvent[iter.Seq[types.NamespacedName]]
+}
+
+var _ reconcile.Reconciler = (*TrainingRuntimeReconciler)(nil)
+var _ TrainJobWatcher = (*TrainingRuntimeReconciler)(nil)
+
+func NewTrainingRuntimeReconciler(cli client.Client, recorder record.EventRecorder) *TrainingRuntimeReconciler {
+	return &TrainingRuntimeReconciler{
+		log:                      ctrl.Log.WithName("trainingruntime-controller"),
+		client:                   cli,
+		recorder:                 recorder,
+		nonRuntimeObjectUpdateCh: make(chan event.TypedGenericEvent[iter.Seq[types.NamespacedName]], updateChBuffer),
+	}
+}
+
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainingruntimes,verbs=create;get;list;watch;update;patch;delete
```

Comment:
Practically, it doesn't need the `create` nor `delete` verbs right?

Reply:
@tenzen-y Btw, can we just grant permission for trainingruntimes/finalizers ?

Reply:
https://github.com/kubeflow/trainer/pull/2608#discussion_r2068340963

Reply:
Done.